### ISSUE TITLE
victim of piracy is not getting a message/info about this fact

### ISSUE THREAD

**Issue by LynxAbraxas** [civctp2/civctp2#75]
In relation to #49 and PR #71, it seems that the victim of piracy is not getting a message about this fact, if playing with the ctp2 style of pirating rule. (This only seems to happen if ctp1 style of pirating rule is used.)
Additionally, as I understand the code, the trade gold of the route that is getting pirated is not removed from the victim (only additionally added to the attacker).

**Comment by LynxAbraxas** [comment]
I have the impression that the pirating code is not at an ideal position possibly wrong there:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1650-L1681
There is also InterceptTrade:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L5862
I noticed that pirating only succeeds sometimes, but I cannot find any randomization in the code concerning success.
Additionally, I get a message if I pirate a route from an AI I'm not at war with, but when the AI pirates a route of mine, I do not get a message (#76 should amend that) nor is the pirating listed in the column of the trade window. 

@MartinGuehmann what do you think about the pirating code?
Do you see any randomization concerning its success?
Why is it different to that of e.g. a cleric selling indulgences (https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L4330-L4428)?
Is the code missing the subtraction of the pirated gold from the owner of the route?

**Comment by MartinGuehmann** [comment]
@LynxAbraxas, the code is a bit odd, I would put it in the ArmyBeginTurn event, but rather at the code when the prirating is happining, it could be in UnitData::InterceptTrade() or in the GEV_SetPiratingArmy event. I know these events are a bit tricky, first you have to track down the code and you have to keep in mind that the events are executed asynchronously. For instance GEV_SetPiratingArmy is added after current, but it will only executed after current directly if nothing afterwards has been added after current.

GEV_SetPiratingArmy leads you to tradeevent.cpp, and there it just sets the pirating army. Now you have to track what is done if there is a pirating army.

**Comment by LynxAbraxas** [comment]
Thanks @MartinGuehmann  for taking a look.
> Now you have to track what is done if there is a pirating army.

And that leads me back to:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1658
without any randomization in between, so I guess there is a bug there.
Oddly, I have a save-game where an AI player has a boat that comes up with a pirate flag on one of my trade routes (when that player is evaluated between turns). So, I assume that comes from here:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L5917
Or is there any other special action that leads to a pirate flag special effect?
However, I neither get pirating listed in the trade manager
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/ui/interface/trademanager.cpp#L1013-L1032
nor any option in the diplomacy manager for asking to stop piracy nor a message with #76. So I really wonder if the processing of pirating is lost there and if so where. Possibly that is the same source for the randomness that I experience when pirating. 
Could it be that the event is lost in some cases?

**Comment by MartinGuehmann** [comment]
> > Now you have to track what is done if there is a pirating army.
> 
> And that leads me back to:
> [civctp2/ctp2_code/gs/gameobj/ArmyData.cpp](https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1658)
> 

Well, I rather meant how is then the income from that route calculated. So that would be rather some code in CityData.cpp or Player.cpp.

**Comment by LynxAbraxas** [comment]
> how is then the income from that route calculated

As I understand, it comes from:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1659
`route->GetValue()` is what is shown in the trade manager and `PIRACY_WASTE_COEFFICIENT` seems constant:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_data/default/gamedata/Const.txt#L370
Anyway, for clearification: With randomness I am meaning a boolean randomness in whether pirating succeeds or fails, not a randomness in the amount of gold captured; i.e. what randomness leads to no pirating messages for aggressor or victim, no listing in the trade manager and the absence of an option for diplomatic negotiations to ask for stopping pirating.

**Comment by LynxAbraxas** [comment]
Tried to debug this a bit more. I have the impression that the problem lies here:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1658
This condition does not seem to be true as often as it should be.
The pirating army seems to be set here:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/UnitData.cpp#L2499-L2504
I find it odd, that it seems to be handled by the army while being set at the unit level.

Also, with https://github.com/civctp2/civctp2/pull/76/commits/e866a508830222e2a0b9e525caafc54f12e300d3 (of #76) compiled in, I only get the message of being pirated (as well as the indication in the trade manager) if I switch players that do the pirating and getting pirated after one turn. So I currently suspect that some of the code is dependent on the player being human or AI, which should not be the case.
@MartinGuehmann Do you have an idea what could be wrong? I cannot see what is wrong in this case.

**Comment by LynxAbraxas** [comment]
I also have the impression that this code:
https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/UnitData.cpp#L2508
introduces CTP1 pirating behaviour independent on what was chosen as a game rule when starting the game. 
I have observed that a trade route of mine was destroyed by the AI pirating it for some rounds even though I chose CTP2 trade rules at the beginning. Do the rule I choose only apply for the human player and might differ for an AI player or would that be a bug?

**Comment by MartinGuehmann** [comment]
> I also have the impression that this code:
> [civctp2/ctp2_code/gs/gameobj/UnitData.cpp](https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/UnitData.cpp#L2508)
> 
> Line 2508 in [e0edb94](/civctp2/civctp2/commit/e0edb94b97916c979857a712ab8530779f9055b4)
> 
>  #ifdef CTP1_TRADE 
> 
> introduces CTP1 pirating behaviour independent on what was chosen as a game rule when starting the game.
> I have observed that a trade route of mine was destroyed by the AI pirating it for some rounds even though I chose CTP2 trade rules at the beginning. Do the rule I choose only apply for the human player and might differ for an AI player or would that be a bug?

Actually, makes CTP2 behave as CTP1, but for that you have to compile the code by defining CTP1_TRADE somewhere, actually I think it has also some other side effects.

Actually, in CTP1 the trade route is destroyed when it is pirated, that is exactly what the code guarded by the #ifdef does. It also sends a message to the victim, it also gives the pirating civ if the pirating army was visible to the victim.

I don't know why your trade route was destroyed, if it were an AI trade route I would assume that it would just abandon it. Maybe the trade routes get killed automatically, if they do not generate any profit anymore. Who knows.

**Comment by MartinGuehmann** [comment]
> Tried to debug this a bit more. I have the impression that the problem lies here:
> [civctp2/ctp2_code/gs/gameobj/ArmyData.cpp](https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/ArmyData.cpp#L1658)
> 
> Line 1658 in [e0edb94](/civctp2/civctp2/commit/e0edb94b97916c979857a712ab8530779f9055b4)
> 
>  if(route->GetPiratingArmy().m_id == m_id) { 
> 
> This condition does not seem to be true as often as it should be.
> The pirating army seems to be set here:
> [civctp2/ctp2_code/gs/gameobj/UnitData.cpp](https://github.com/civctp2/civctp2/blob/e0edb94b97916c979857a712ab8530779f9055b4/ctp2_code/gs/gameobj/UnitData.cpp#L2499-L2504)
> 

Well, if another player pirates the trade during his turn which comes after your turn then all the pirate income would go to his pockets. In principle, you could use the debug version and use DPRINTF, to see what the army IDs are.

> Lines 2499 to 2504 in [e0edb94](/civctp2/civctp2/commit/e0edb94b97916c979857a712ab8530779f9055b4)
> 
>  g_gevManager->AddEvent 
>      (GEV_INSERT_AfterCurrent,   GEV_SetPiratingArmy, 
>  	 GEA_TradeRoute,            route, 
>  	 GEA_Army,                  m_army, 
>  	 GEA_End 
>      ); 
> 
> I find it odd, that it seems to be handled by the army while being set at the unit level.

Well, that should be fine, since it is only done once per army, even so it is not so clean.

> Also, with [e866a50](https://github.com/civctp2/civctp2/commit/e866a508830222e2a0b9e525caafc54f12e300d3) (of #76) compiled in, I only get the message of being pirated (as well as the indication in the trade manager) if I switch players that do the pirating and getting pirated after one turn. So I currently suspect that some of the code is dependent on the player being human or AI, which should not be the case.
> @MartinGuehmann Do you have an idea what could be wrong? I cannot see what is wrong in this case.

So when you switch players, did you attach the player on screen? Remember if you switch via the cheat editor than the player on screen is human, that is also true if you just open the cheat editor, even so the current player on screen was an AI.

The code should say something like IsHuman or IsRobot, for code only specific to the AI. Note that !IsHuman != IsRobot, even so sometimes it is used like that. There is a third player type, which is network.

**Comment by LynxAbraxas** [comment]
> So when you switch players, did you attach the player on screen?

Just switched with the cheat window (empire tab) first player 1 then player 2 then back to 1 hit "end turn". That is what makes me wonder if some code special to a human player is causing the problems
Will try to create a reproducible minimal save-game for this.

**Comment by MartinGuehmann** [comment]
> Just switched with the cheat window (empire tab) first player 1 then player 2 then back to 1 hit "end turn". 

Well, that shouldn't change anything, unless your player 1 was an AI before.

**Comment by LynxAbraxas** [comment]
In regard to the insights from https://github.com/civctp2/civctp2/pull/76#issuecomment-482749127

Here is a [save-game](https://github.com/civctp2/civctp2/files/3077663/save-game.zip) with at least locally reproducible results with these instructions:


trade route got pirated message:

1. pirate as Persian
2. change to player 2 (Crete, Minoans) in cheat window, close cheat window! -> F4 shows Persian pirate, F7 shows no anti-piracy option (Minoans fire trireme has already pirated)
3. change back to player 1, end turn as Persian
4. Msg: trade route got pirated message is shown, but F4 shows no piracy and F7 no anti-piracy option

diplomacy window (F7) show anti-piracy option for other player:
1. pirate as Persian
2. end turn (as Persian)
3. pirate again as Persian
4. change to player 2 -> F4 shows no piracy BUT F7 shows anti-piracy option

diplomacy window (F7) show anti-piracy option for first player:
1. pirate as Persian
2. change to player 2 (Crete, Minoans) in cheat window, close cheat window! -> F4 shows Persian pirate, F7 shows no anti-piracy option (Minoans fire trireme has already pirated)
3. end turn as player 2 -> Msg: trade route got pirated message is shown
4. pirate Persian route again
5. end turn as player 2
6. change to player 1 ->  F4 shows no piracy BUT F7 shows anti-piracy option

Variant to the one above for first player, that should work (but does not):
1. pirate as Persian
2. change to player 2 (Crete, Minoans) in cheat window, close cheat window! -> F4 shows Persian pirate, F7 shows no anti-piracy option (Minoans fire trireme has already pirated)
3. end turn as player 2 -> Msg: trade route got pirated message is shown
4. pirate Persian route again
5. change to player 1
6. end turn as player 1 -> Msg: trade route got pirated message is shown BUT F4 shows no piracy and F7 shows no anti-piracy option
 

- So it seems a route needs to be pirated twice in a row by the same player for the anti-piracy option to show up under F7
- F4 shows piracy only after changing player and piracy happend just before and before the turn is ended
- trade route got pirated message only shows up after changing player and piracy happend just before and turn was ended

@MartinGuehmann can you reproduce these results under linux, windows? Any idea where the error could lie?

**Comment by LynxAbraxas** [comment]
In regard to https://github.com/civctp2/civctp2/pull/76#issuecomment-482749127:
> So if two players are pirating the same trade route over turns then nobody gets the bounty. Of course, still the trade route does not give any gold to the owner.

It would be great to change that. One possible way I can imagine to reasonably choose one of the concurring players to get the bounty would be to choose the unit/army that is closest to the source of the trade route. That I would find logical, since then if that unit catches the bounty there is no trade to catch for those that follow along the route, and this would introduce other strategic positions instead of just pirating directly before the destination where often many routes join.
@MartinGuehmann  do you know if there is already some code in ctp2 that determines which unit is closer to the source of a route (possibly for other purposes, like AI goals)?

**Comment by MartinGuehmann** [comment]
> In regard to [#76 (comment)](https://github.com/civctp2/civctp2/pull/76#issuecomment-482749127):
> 
> > So if two players are pirating the same trade route over turns then nobody gets the bounty. Of course, still the trade route does not give any gold to the owner.
> 
> It would be great to change that. One possible way I can imagine to reasonably choose one of the concurring players to get the bounty would be to choose the unit/army that is closest to the source of the trade route. That I would find logical, since then if that unit catches the bounty there is no trade to catch for those that follow along the route, and this would introduce other strategic positions instead of just pirating directly before the destination where often many routes join.

Well, I would prefer to give it to the one who pirated the route first. I find it quite logical, if you wanna rob somebody and he has been already robbed then nothing is left for you to rob.

> @MartinGuehmann do you know if there is already some code in ctp2 that determines which unit is closer to the source of a route (possibly for other purposes, like AI goals)?

There are a few Distance measurements in the code, I think I would go for the SquaredDistance.

**Comment by LynxAbraxas** [comment]
> Well, I would prefer to give it to the one who pirated the route first. I find it quite logical, if you wanna rob somebody and he has been already robbed then nothing is left for you to rob.

True, me too, but how to determine who is the first if there are multiple pirating civs on the same route? Just because the player number is lower would be unfair, wouldn't it?

To my understanding each player (human or AI) has his/her own slot, let's call it sub-turn, where the interaction takes place. These sub-turns are ordered according to the player number, barbarians come last and then the actual turn (for all) ends and a new turn starts, in which player 1 is always the first to start with his/her new sub-turn.
Please tell, me if that understanding is not correct. If it is, then in the current state the last player has the advantage that his army/unit ID is set last and will get the bounty if in the following turn and all the sub-turns before the last player no pirating happens again.
That's why I think the distance to the source would be an appropriate and a fair way, and one that can be determined in the general scope of ctp2.
Otherwise you could say that the first who catches the good icon as it is transported along the route gets the bounty but that would lead far away from the turn-based game and would probably hard to implement in the current scope.

**Comment by LynxAbraxas** [comment]
> There are a few Distance measurements in the code, I think I would go for the SquaredDistance.

Well, I guess, the most fair/appropriate would be the actual distance along the trade route as it can be very curved.

**Comment by LynxAbraxas** [comment]
I think the discussion has brought up another issue, the one to discuss here is that the victim does not get the info neither that from #76 like here (from the procedures described in https://github.com/civctp2/civctp2/issues/75#issuecomment-483026718):

![ss_2019-04-11_18:51:53](https://user-images.githubusercontent.com/21012234/56098047-54dc3d00-5efc-11e9-881c-adc46e4cfc7d.png)


 nor the colored square in the trade window under the summary like:

![ss_2019-04-11_18:51:12](https://user-images.githubusercontent.com/21012234/56098033-41c96d00-5efc-11e9-829b-01103b6f8880.png)

**Comment by MartinGuehmann** [comment]
> > Well, I would prefer to give it to the one who pirated the route first. I find it quite logical, if you wanna rob somebody and he has been already robbed then nothing is left for you to rob.
> 
> True, me too, but how to determine who is the first if there are multiple pirating civs on the same route? Just because the player number is lower would be unfair, wouldn't it?

In the CTP1 setting the trade route is killed. And the pirate gets the loot or booty. (Sorry I should have looked it up. Bounty is not exactly that, rather some money you get when you catch the pirate and hand him over to the authorities. But actually, the more I look it up, I get the feeling that I actually could stretch the meaning so far.)

Anyway, the trade route is pirated for one turn, so whoever comes first, robs first. Usually, the two pirates won't arrive at the same turn, so that will be fine. And if it is indeed the same turn, well it is a turn based game. Usually, you have on board games a random component, but it is still who comes first is served first.

> To my understanding each player (human or AI) has his/her own slot, let's call it sub-turn, where the interaction takes place. These sub-turns are ordered according to the player number, barbarians come last and then the actual turn (for all) ends and a new turn starts, in which player 1 is always the first to start with his/her new sub-turn.

Not exactly, if a new civ is created then it has its turn after the player's turn when it was created. But well, for the argument it does not matter.

> Please tell, me if that understanding is not correct. If it is, then in the current state the last player has the advantage that his army/unit ID is set last and will get the bounty if in the following turn and all the sub-turns before the last player no pirating happens again.

Yes, that it is how it is, unless everybody pirates again. Then nobody gets anything.

> That's why I think the distance to the source would be an appropriate and a fair way, and one that can be determined in the general scope of ctp2.

Yes, in fact you only have to count the nodes of the path, I think I have done something like this with the empire center calculation.

> Otherwise you could say that the first who catches the good icon as it is transported along the route gets the bounty but that would lead far away from the turn-based game and would probably hard to implement in the current scope.

I don't see why, you just get closer to CTP1, in principle you give the booty at the time when you pirate, and if the trade route already has been pirated then you get nothing.

**Comment by LynxAbraxas** [comment]
Thanks to @MartinGuehmann it is at least resovled that the trade gold is *not added*, see https://github.com/civctp2/civctp2/pull/76#issuecomment-482749127
and
https://github.com/civctp2/civctp2/blob/b3ac27230a0cfba4cba2c912c81ab5abf18fa3ae/ctp2_code/gs/gameobj/CityData.cpp#L4433-L4461

So let's keep this issue to track the fact that a victim of piracy neither gets a message nor any other kind of info about being subject to piracy, like in the trade manager (as it should normally be the case):

![ss_2019-04-11_18:51:12](https://user-images.githubusercontent.com/21012234/56098033-41c96d00-5efc-11e9-829b-01103b6f8880.png)

#117 is for discussion how to handle the case that multiple civs pirate the same route and
#116 is for the relation to diplomacy, which could in principle even handle multiple pirating civs.

While the trade manager info in its current design can only handle giving info on one pirating civ, messages like #76 could be sent for multiple pirating civs (even on the same route).

**Comment by MartinGuehmann** [comment]
> So let's keep this issue to track the fact that a victim of piracy neither gets a message nor any other kind of info about being subject to piracy, like in the trade manager (as it should normally be the case):

Actually, you get some info about being pirated: In the trade manage you have the column with the pirate flag and orange square. That should say that the Greeks are pirating your trade route to the Greek city of Knossos, which is quite, since the Greek are involved in the trade, they should consider it as their own route and thus should not pirate it.

**Comment by LynxAbraxas** [comment]
> Actually, you get some info about being pirated: In the trade manage you have the column with the pirate flag and orange square. That should say that the Greeks are pirating your trade route to the Greek city of Knossos, which is quite, since the Greek are involved in the trade, they should consider it as their own route and thus should not pirate it.

That image comes from the testing described above where I only got that colored square by the particular switching described above. Part of THIS issue is that this colored square never appears under normal circumstances, i.e. where you do not switch players in the cheat mode.

Despite that, your comment makes me wonder if the receiving civ of a good has any benefits from the trade? I know that in civ3 you can get access to a strategic good (e.g. oil for being able to build tanks) by asking another civ to sell/trade it to you. However, as far as I can remember, such strategic good trading is only mentioned in comments but has not been realised so far. So are there any benefits fot the receiving civ in the current state  of civctp2?

**Comment by MartinGuehmann** [comment]
> Despite that, your comment makes me wonder if the receiving civ of a good has any benefits from the trade? I know that in civ3 you can get access to a strategic good (e.g. oil for being able to build tanks) by asking another civ to sell/trade it to you. However, as far as I can remember, such strategic good trading is only mentioned in comments but has not been realised so far. So are there any benefits fot the receiving civ in the current state of civctp2?

Of course, both partners in trade should benefit from it. As far as I know you get the income from the trade and if you buy from another civ you pay for that good. Of course that could be more. Well, for my longer test games that are rather AI only games and trade is not really the main focus so far. It is rather getting the AI to conquer nicely and using its troops.

**Comment by LynxAbraxas** [comment]
> Of course, both partners in trade should benefit from it. 

That would be naturally expected and a great concept but I do not think that this is currently the case.

> if you buy from another civ you pay for that good

Hm, I came across this concept somewhere (possibly only in code comments) but never found a way of doing that. Do you know how to actually buy a good?

> AI only games and trade is not really the main focus so far

Odd, in all my games as soon as trade is available to AI civs they do start trading heavily and also the AI uses its troops (mostly ships) to pirate others routes.
Actually, I have the impression that the AI also "parks" transport-ships on routes possibly for pirating, just it does not notice/learn/was not tought that a transporter cannot pirate, could this be a bug?

**Comment by MartinGuehmann** [comment]
> > Of course, both partners in trade should benefit from it.
> 
> That would be naturally expected and a great concept but I do not think that this is currently the case.

Well, I'd say currently it is hard to see whether it is so.

> > if you buy from another civ you pay for that good
> 
> Hm, I came across this concept somewhere (possibly only in code comments) but never found a way of doing that. Do you know how to actually buy a good?

That you pay for the good if you buy it from somewhere is in the code. And for buying you go the trade manager click on friendly or all and then you see goods from other civs that you can buy. Of course, that is a bit difficult if you are on war with about like my current AI. Well, actually, on the screenshots you created just such a trade route between Knossos and Persepolis.

> > AI only games and trade is not really the main focus so far
> 
> Odd, in all my games as soon as trade is available to AI civs they do start trading heavily and also the AI uses its troops (mostly ships) to pirate others routes.
> Actually, I have the impression that the AI also "parks" transport-ships on routes possibly for pirating, just it does not notice/learn/was not tought that a transporter cannot pirate, could this be a bug?

Well, yes that what is what I see, too. I although think that the AI is pirating the trade routes from transporters. Well, if it does so then it is the same problem as executing other special orders from a ship.

Well, as I said for AI testing that is not the main concern. I rather look at it as a by the way thing, since so far it was more important to make the AI attack and conquer, and also settle at the right places. As far as I could see the AI was pirating and had no problem with it. Actually, this issue isn't really about the AI. In an AI only game you don't get any message, whether a trade route was pirated or whatever. Only the Diplo window pops up from time to time and if your spies investigate a city you get the window. But that is.

**Comment by LynxAbraxas** [comment]
> And for buying you go the trade manager click on friendly or all and then you see goods from other civs that you can buy.

As I understand it, that is selling to another civ (which brings more gold than trading with your own cities) not buying from another civ. These routes carry a good of a city of mine to another civ's city. By buying I would expect to initiate trade in the other way round, getting another civ to create a trade route of a good to one of my cities. However, without any benefits from receiving a good there seems to be no point in this so far, until we might have some strategic goods implemented that owning becomes a requisite to e.g. build tanks or atomic bombs.

**Comment by LynxAbraxas** [comment]
> As far as I could see the AI was pirating and had no problem with it.

Not a "problem" but possibly no benefit. 
The message/info for the victim, i.e. what this issue means to point at, is of course for a human player. Currently the AI seems to get this info and reacts with a diplomatic proposal asking you not to pirate their trade routes. However in the current state, a human player gets neither a message nor the color in the pirate column nor the diplomatic option in the diplomacy window to ask for "stop pirating", which is probably a bug otherwise kind of unfair and in the end frustrating since you have no handle to tell the AI to stop. The only way is to hunt pirating units, declaring war (without being able to say why) and to destroy the pirating unit/army, while the pirating animation is not visible within "fog of war " or in unexpored tiles.

**Comment by LynxAbraxas** [comment]
> These routes carry a good of a city of mine to another civ's city. By buying I would expect to initiate trade in the other way round, getting another civ to create a trade route of a good to one of my cities.

Seems there is some code concerning actual buying
https://github.com/civctp2/civctp2/blob/b3ac27230a0cfba4cba2c912c81ab5abf18fa3ae/ctp2_code/gs/gameobj/CityData.cpp#L4766-L4783
However I've no idea how to get there from within the game. @MartinGuehmann any ideas how?

**Comment by LynxAbraxas** [comment]
@MartinGuehmann Do you  have any idea why `StopPirating();` is called in `ArmyData::AutoAddOrders`?
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L6057
It is there since the first commit so I would assume that it came from activision.
Do you know if this issue here already applied to the original version of ctp2 (i.e. no colored square and probably also no diplomacy option)?

**Comment by LynxAbraxas** [comment]
The coloring of the square seems directly to depend on the evaluation of `IsBeingPirated()`
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/ui/interface/trademanager.cpp#L1028-L1030
(not sure when one would end up in [`TradeManager::CompareSummaryItems`](https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/ui/interface/trademanager.cpp#L938-L951))

Also the code that determins if the value of a trade route is added as gold for the owner (if not pirated, and thereby also the message offered with #125):
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/CityData.cpp#L4440-L4447
as well as `ComputeTradeRoutePiracyRisk`:
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/ai/diplomacy/diplomat.cpp#L4706

 `IsBeingPirated()` comes from:
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/TradeRouteData.cpp#L628-L631
and depends on the pirating army that is set with:
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/TradeRouteData.cpp#L618-L621

My impression from trying to debug this is that the setting of the armies of different players somehow interfers such that when  `IsBeingPirated()` is called in the cases above `IsValid()`
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/Army.cpp#L45-L48
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ObjPool.h#L138-L150
from the set army returns false when another player is evaluated even though the pirating flag action animation is visible during the evaluation of the other player sub-turns.

@MartinGuehmann Can you help me with resolving this issue? I've spent multiple days now without any success of finding the source for the issue and am far from solving this.

**Comment by LynxAbraxas** [comment]
@MartinGuehmann is `g_theArmyPool`
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/Army.cpp#L47
a pool of all armies independent of the current evaluated player?

Under what circumstances would 
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/Army.cpp#L45
not return `true`? Army destroyed?
Or the other way round should an army that is visible on the map have a return value of `true`?
Would it make a difference if it is just a single unit?

**Comment by LynxAbraxas** [comment]
Finally some directions: It seems that disabling `StopPirating` in `AutoAddOrders`,  `AutoAddOrdersWrongTurn` and `AddOrders` (where purpose is not clear) as in https://github.com/civctp2/civctp2/pull/125/commits/5201be5f1c31c5b00116dc8267e8a8331f7384a3 at least partially solves the problem, then the square is colored, the message appears (actually multiple times) and the diplomacy option is there!
So far it is not clear to my why that has this drastic effect, any ideas @MartinGuehmann?

**Comment by LynxAbraxas** [comment]
It seems that `StopPirating` in `AddOrders`
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L6204
is (mostly) the culprit. It seems to get called by the AI (probably) every turn, which is disabling the mechanism to check if a trade route is pirated. However, just removing it there leads to the problem, that  `StopPirating` is never called, not even if moving the unit/army off the route it was pirating. Moving off leads to `piratedByMe == 0`
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L1703-L1705
which calls `StopPirating` but there
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L10959-L10975
(even though `m_isPirating` is set to false in [any case](https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L10978))
the new `m_pos` does not have the old trade route (as we consider moving off the route, not along) and therefore does not and cannot set the pirating army of the former trade route(s) to zero:
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L10971
which in turn is needed to make `IsBeingPirated()` to return `false`.

So despite that moving off removes piracy for the attacker, moving off does not lead to the removal of the color in the square nor the diplomacy entry nor the message of #125. Furthermore, it "appears" to the AI as if the route(s) are constantly pirated and so it cancels the routes according to its rules (after 3 turns? @MartinGuehmann where did you find that condition?).

This is of course not OK, so we need to find a more sophisticated solution. As the removal seems to have no other bad effect, I am considering adding  `StopPirating` to a function that is called before a unit is moved. @MartinGuehmann do you know if there is already such a function where this could be placed appropriately? Or do you have any other ideas?

**Comment by MartinGuehmann** [comment]
> Finally some directions: It seems that disabling `StopPirating` in `AutoAddOrders`, `AutoAddOrdersWrongTurn` and `AddOrders` (where purpose is not clear) as in [5201be5](https://github.com/civctp2/civctp2/commit/5201be5f1c31c5b00116dc8267e8a8331f7384a3) at least partially solves the problem, then the square is colored, the message appears (actually multiple times) and the diplomacy option is there!
> So far it is not clear to my why that has this drastic effect, any ideas @MartinGuehmann?

That's now quite a lot. If you give new orders then you the army is supposed to do something else, and that would also be something else then pirating. So from that point of view it is logical to stop pirating. Now at least for AIs, you have the scheduler and the process match cycles. The AI has five or six cycles where it gives all the orders and after each cycle it moves the armies. That's because it uses the slic events for this, which work asynchronously.

So for instance in the first cycle it moves all the units for city siege to the meeting point (rally point). In the next cycle it groups all the units at the meeting point and ideally in the third cycle the units have been grouped and can now go for sieging the target city.

In case of pirating, we may already have the army at the target, so it pirates in the first cycle, the AI will get to the same conclusion in the second cycle and in every cycle following. So you will get then the message 5 or 6 times per turn. So StopPirating should only be called if the order is not OrderPirate or if it is OrderPirate but not the same trade route.

**Comment by LynxAbraxas** [comment]
Many thanks @MartinGuehmann  for your feedback. 
> So StopPirating should only be called if the order is not OrderPirate or if it is OrderPirate but not the same trade route.

How would you suggest to realize this? I'm not familiar with the AI code. Can you help me with a commit?

**Comment by LynxAbraxas** [comment]
As far as I can imagine, only moving off the route is a critical action/order, but combining it with moving off to another tile (possibly on the same route(s)) would probably be OK. Is there a specific function for that where `StopPirating` would fit in?

**Comment by MartinGuehmann** [comment]
> is (mostly) the culprit. It seems to get called by the AI (probably) every turn, which is disabling the mechanism to check if a trade route is pirated. 

Well, the AI pirates the route every turn, so that could disable it.

> So despite that moving off removes piracy for the attacker, moving off does not lead to the removal of the color in the square nor the diplomacy entry nor the message of #125. Furthermore, it "appears" to the AI as if the route(s) are constantly pirated and so it cancels the routes according to its rules (after 3 turns? @MartinGuehmann where did you find that condition?).

That is rather an assumption, I know that the AI cancels routes that are not profitable. For instance if you load a savegame with less goods in the database and the game cannot convert the goods properly then you end up with all ruby trade routes, which the AI cancels after a turn.

> 
> This is of course not OK, so we need to find a more sophisticated solution. As the removal seems to have no other bad effect, I am considering adding `StopPirating` to a function that is called before a unit is moved. @MartinGuehmann do you know if there is already such a function where this could be placed appropriately? Or do you have any other ideas?

Well, you could put it into MoveUnits, but still if you are next to a city and can bombard it, and decide bombarding it then you should stop pirating if you are doing it.

**Comment by MartinGuehmann** [comment]
> > So StopPirating should only be called if the order is not OrderPirate or if it is OrderPirate but not the same trade route.
> 
> How would you suggest to realize this? I'm not familiar with the AI code. Can you help me with a commit?

That's not AI code, you would put that into the code that calls StopPirating.

**Comment by LynxAbraxas** [comment]
> Well, you could put it into MoveUnits

So how about just before
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L7910
Would that be appropriate? Important is that `m_pos` has not been changed before `StopPirating` is called.

**Comment by LynxAbraxas** [comment]
>That's not AI code, you would put that into the code that calls StopPirating.

OK, so basically into `AddOrder`? However, I cannot find a `UNIT_ORDER` for pirating:
https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/UnitTypes.h#L121-L197

**Comment by LynxAbraxas** [comment]
NB: Are all these orders currently available from within the game? How would one isssue e.g.  `UNIT_ORDER_PATROL`  or `UNIT_ORDER_CIRCULAR_PATROL` from within the game?

**Comment by MartinGuehmann** [comment]
> > Well, you could put it into MoveUnits
> 
> So how about just before
> https://github.com/civctp2/civctp2/blob/0668dd199c2519ff8ab7d2871e7c7a89e21c8e52/ctp2_code/gs/gameobj/ArmyData.cpp#L7910
> 
> 
> Would that be appropriate? Important is that `m_pos` has not been changed before `StopPirating` is called.

Not really. Let's see to get some code around StopPirating. Should be real code, but I cannot guarantee that it works out of the box.

```
if(point != m_pos                         // I hope point is the target position
|| order != UNIT_ORDER_INTERCEPT_TRADE
  ) 
{
	StopPirating();
}
```

At least you came up with the list of orders. I should now have picked the right one.

**Comment by LynxAbraxas** [comment]
@MartinGuehmann Thanks for pointing me to `UNIT_ORDER_INTERCEPT_TRADE` (I must be getting tired...).

**Comment by LynxAbraxas** [comment]
Let's see if https://github.com/civctp2/civctp2/pull/125/commits/cf9595ad7fb57672a22f2cbd2ab4ceaa33640961 solves this.

**Comment by MartinGuehmann** [comment]
> NB: Are all these orders currently available from within the game? How would one isssue e.g. `UNIT_ORDER_PATROL` or `UNIT_ORDER_CIRCULAR_PATROL` from within the game?

I would say not at all, actually for those you have also some AI goals, which however the AI does not use, because the number it is supposed to evaluate is zero.

Some are for internal use for instance: UNIT_ORDER_TELEPORT_TO is used if a cargo heli unloads it cargo.

Then we have UNIT_ORDER_UNLOAD_SELECTED_STACK, which you use implicitly, you can go to the cargo list and select the units you wanna unload.

And of course, something could be just an idea that was never implemented.

**Comment by LynxAbraxas** [comment]
There is so much new stuff to find in the code, just amazing what concepts and complexity have already been implemented (mostly). Just came accross one while debugging (#126) which seems to have been disabled (possibly unintended).

**Comment by MartinGuehmann** [comment]
> There is so much new stuff to find in the code, just amazing what concepts and complexity have already been implemented (mostly). Just came accross one while debugging (#126) which seems to have been disabled (possibly unintended).

Which one do you mean? In fact, you enabled three things. And the most plausible reason that they have been disabled is that those features haven't been finished.

You know, E had many nice ideas, but he wasn't a programmer, that makes the quality of code bad. For instance, he copied a pointer and assumed he copied the object itself. And the other thing is that he did not test all these things. So you cannot be sure that the things he added do work. For instance we have the possibility to disable the city limit. However, in its current form at best it is just partial, even so it would make sense to drop the city limit, entirely.

Also the build item and the turns to the next pop indicators at the city names were his idea. I don't remember whether I actually fixed these things. But of course those are nice.

**Comment by LynxAbraxas** [comment]
Solved with #127.

**Comment by LynxAbraxas** [comment]
From https://github.com/civctp2/civctp2/issues/75#issuecomment-488127228
> constantly pirated and so it cancels the routes according to its rules (after 3 turns?

This seems to come from
https://github.com/civctp2/civctp2/blob/2c9b619e1bb4fe4e3fab1b86dddf521482e930d7/ctp2_data/default/aidata/strategies.txt#L503-L504
as used here
https://github.com/civctp2/civctp2/blob/2c9b619e1bb4fe4e3fab1b86dddf521482e930d7/ctp2_code/ai/diplomacy/diplomat.cpp#L4447
